gitindex/index.go

```diff
@@ -572,6 +577,37 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
 	return true, builder.Finish()
 }
 
+// openRepo opens a git repository in a way that's optimized for indexing.
+//
+// It copies the relevant logic from git.PlainOpen, and enables the filesystem KeepDescriptors option. This
+// caches the packfile handles, preventing the packfile from being opened then closed on every object access.
+func openRepo(repoDir string) (*git.Repository, io.Closer, error) {
+	fs := osfs.New(repoDir)
+
+	// Check if the root directory exists.
+	if _, err := fs.Stat(""); err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil, git.ErrRepositoryNotExists
+		}
+		return nil, nil, err
+	}
+
+	// If there's a .git directory, use that as the new root.
+	if _, err := fs.Stat(git.GitDirName); err == nil {
+		if fs, err = fs.Chroot(git.GitDirName); err != nil {
+			return nil, nil, fmt.Errorf("fs.Chroot: %w", err)
+		}
+	}
+
+	s := filesystem.NewStorageWithOptions(fs, cache.NewObjectLRUDefault(), filesystem.Options{
+		KeepDescriptors: true,
```

Comment:
Here's the point of the whole change :) This also sets me up to set `LargeObjectThreshold` in a follow-up PR, which could prove even more important.